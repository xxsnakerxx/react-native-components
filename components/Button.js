import React from 'react';

import {
  TouchableOpacity,
  View,
  Text,
  StyleSheet,
} from 'react-native';

/**
 * @typedef {import("react-native").GestureResponderEvent} GestureResponderEvent
 * @typedef {import("react-native").ImageLoadEventData} ImageLoadEventData
 * @typedef {import("react-native").NativeSyntheticEvent<ImageLoadEventData>} OnLoadEvent
 * @typedef {import("react-native").ViewProps} ViewProps
 * @typedef {import("react-native").ViewStyle} ViewStyle
 * @typedef {import("react-native").StyleProp<ViewStyle>} ViewStyleProp
 * @typedef {import("react-native").TextStyle} TextStyle
 * @typedef {import("react-native").StyleProp<TextStyle>} TextStyleProp
 * @typedef {import("react-native").TouchableOpacityProps} TouchableOpacityProps
 * @typedef {import("react").Props} ReactProps
 *
 * @typedef ButtonProps
 * @prop {ViewStyleProp} [containerStyle=null]
 * @prop {TextStyleProp} [textStyle=null]
 * @prop {ViewStyleProp} [disabledContainerStyle=null]
 * @prop {TextStyleProp} [disabledTextStyle=null]
 *
 * @typedef {TouchableOpacityProps & ButtonProps & ReactProps} Props
 */

/**
 * @param {Props} props
 */
const Button = (props) => {
  const {
    containerStyle,
    textStyle,
    disabledContainerStyle,
    disabledTextStyle,
    children,
    ...touchableProps
  } = props;

  const _disabledContainerStyle = touchableProps.disabled ? disabledContainerStyle : null;
  const _disabledTextStyle = touchableProps.disabled ? disabledTextStyle : null;

  const isTextButton = typeof children === 'string';

  return (
    <TouchableOpacity
      testID="Button"
      activeOpacity={!touchableProps.onPress ? 1 : undefined}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...touchableProps}
    >
      {
        isTextButton
          ? (
            <View
              style={[
                containerStyle,
                _disabledContainerStyle,
              ]}
            >
              <Text
                style={[
                  styles.text,
                  textStyle,
                  _disabledTextStyle,
                ]}
              >
                {children}
              </Text>
            </View>
          )
          : (
            <View
              style={[
                containerStyle,
                _disabledContainerStyle,
              ]}
            >
              {children}
            </View>
          )
      }
    </TouchableOpacity>
  );
};

Button.defaultProps = {
  textStyle: null,
  containerStyle: null,
  disabledContainerStyle: null,
  disabledTextStyle: null,
};

export default Button;

const styles = StyleSheet.create({
  text: {
    textAlign: 'center',
  },
});
