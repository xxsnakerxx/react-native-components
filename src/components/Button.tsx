import React from 'react';

import {
  TouchableOpacity,
  TouchableOpacityProps,
  Text,
  StyleSheet,
  StyleProp,
  ViewStyle,
  TextStyle,
} from 'react-native';

interface ButtonProps {
  containerStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  disabledContainerStyle?: StyleProp<ViewStyle>;
  disabledTextStyle?: StyleProp<TextStyle>;
}

export type Props = TouchableOpacityProps & ButtonProps;

const Button: React.FC<Props> = (props) => {
  const {
    containerStyle,
    textStyle,
    disabledContainerStyle,
    disabledTextStyle,
    children,
    ...touchableProps
  } = props;

  const _disabledContainerStyle = touchableProps.disabled
    ? disabledContainerStyle
    : null;
  const _disabledTextStyle = touchableProps.disabled ? disabledTextStyle : null;

  const isTextButton = typeof children === 'string';

  return (
    <TouchableOpacity
      testID="Button"
      activeOpacity={!touchableProps.onPress ? 1 : undefined}
      {...touchableProps}
      style={[containerStyle, _disabledContainerStyle]}>
      {isTextButton ? (
        <Text style={[styles.text, textStyle, _disabledTextStyle]}>
          {children}
        </Text>
      ) : (
        children
      )}
    </TouchableOpacity>
  );
};

export default Button;

const styles = StyleSheet.create({
  text: {
    textAlign: 'center',
  },
});
