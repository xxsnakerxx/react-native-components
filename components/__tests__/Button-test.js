import {
  View,
  Text,
} from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import Button from '../Button';

describe('Button', () => {
  const textStyle = { color: '#000' };
  const containerStyle = { height: 100 };
  const disabledStyle = { opacity: 0.4 };

  it('renders correctly with text', () => {
    const tree = renderer.create(
      <Button
        containerStyle={containerStyle}
        textStyle={textStyle}
        disabledStyle={disabledStyle}
      >
        {'Button text'}
      </Button>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with text and disabled', () => {
    const tree = renderer.create(
      <Button
        disabled
        containerStyle={containerStyle}
        textStyle={textStyle}
        disabledStyle={disabledStyle}
      >
        {'Button text'}
      </Button>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with complex children', () => {
    const tree = renderer.create(
      <Button
        containerStyle={containerStyle}
        textStyle={textStyle}
        disabledStyle={disabledStyle}
      >
        <View>
          <Text>
            Complex children
          </Text>
        </View>
      </Button>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with complex children and disabled', () => {
    const tree = renderer.create(
      <Button
        disabled
        containerStyle={containerStyle}
        textStyle={textStyle}
        disabledStyle={disabledStyle}
      >
        <View>
          <Text>
            Complex children
          </Text>
        </View>
      </Button>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
