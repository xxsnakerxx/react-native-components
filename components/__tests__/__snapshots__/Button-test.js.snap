// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly with complex children 1`] = `
<View
  accessible={true}
  focusable={false}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        null,
        Object {
          "height": 100,
        },
      ]
    }
  >
    <View>
      <Text>
        Complex children
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button renders correctly with complex children and disabled 1`] = `
<View
  accessible={true}
  focusable={false}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "opacity": 0.4,
        },
        Object {
          "height": 100,
        },
      ]
    }
  >
    <View>
      <Text>
        Complex children
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button renders correctly with text 1`] = `
<View
  accessible={true}
  focusable={false}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "height": 100,
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        Object {
          "color": "#000",
        },
        null,
        null,
      ]
    }
  >
    Button text
  </Text>
</View>
`;

exports[`Button renders correctly with text and disabled 1`] = `
<View
  accessible={true}
  focusable={false}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "height": 100,
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        Object {
          "color": "#000",
        },
        Object {
          "opacity": 0.4,
        },
        null,
      ]
    }
  >
    Button text
  </Text>
</View>
`;
